producer/writer/message_writer.go

```diff
@@ -279,27 +290,41 @@ func (w *messageWriterImpl) retryUnacknowledged() {
 	var (
 		toBeRetried []*message
 		beforeRetry = w.nowFn()
+		batchSize   = w.opts.MessageRetryBatchSize()
 	)
 	for e != nil {
-		now := w.nowFn()
-		nowNanos := now.UnixNano()
+		beforeBatch := w.nowFn()
+		beforeBatchNanos := beforeBatch.UnixNano()
 		w.Lock()
-		e, toBeRetried = w.retryBatchWithLock(e, nowNanos)
+		e, toBeRetried = w.retryBatchWithLock(e, beforeBatchNanos, batchSize)
 		consumerWriters := w.consumerWriters
```

Comment:
nit: do you need to use a local variable here?

Reply:
yeah I need to make a copy of the list with the lock so I can write to them without lock, I create a new list whenever I update the list